### DISCUSSION TITLE
Collision with walls

### DISCUSSION THREAD

**Discussion by stingerhorse** [AlmasB/FXGL#1164] [Q&A]
Hi there!
I'm currently working on a roguelike game but I can't get any further with the player colliding with walls. I've created the wall as an entity and two variables which have always the value of the current x and y position of the player. If the player collides with the wall, he would be set back to his previous coordinates. The collision gets detected, but I wasn't able to figure out how to give the coordinates to a variable. I think I have to proceed in a completely different way.
Thanks for any advice!

**Comment by AlmasB** [comment]
Hi @stingerhorse 

Have a look at [this example](https://github.com/AlmasB/FXGLGames/blob/d00f131f6911e5e865d540886bfab138167ca004/BattleTanks/src/main/java/com/almasb/fxglgames/tanks/components/MoveComponent.java#L99). I think it does something similar to what you are trying to achieve. Let me know if it works.

**Reply by stingerhorse** [reply]
Thanks for the fast reply. I think the thing with the variables which I described is a bit too complicated. I now tried it to code it similar to what you did in this tutorial: https://www.youtube.com/watch?v=37wfF9GW1vQ&t . Meaning, working more with the types of objects in tiled. The only issue I ran into is the "Entities" class you used in 12:56. The class is not present in my case. The imports are correct.

Really appreciate your help!

**Reply by fleggs** [reply]
Yeah I have the same problem with "Entities". Additionally, I coded it almost exactly as in the video, with the Entities in a new class which implements "EntityFactory", but if I run it, the game crashes and says that no EntityFactory was called.
Thanks in advance.

**Reply by AlmasB** [reply]
`Entities` is a bit of legacy API. It no longer exists in the latest version. Here's the latest version of [the platformer example](https://github.com/AlmasB/FXGLGames/tree/master/Mario).

To add an entity factory, you can call:

```
getGameWorld().addEntityFactory( // your entity factory object)
```

**Reply by fleggs** [reply]
Thanks a lot, I've come a little further. I still struggle with the walls or "platform" in the platformer game. The player just runs right through it. I think I'm missing something where I describe what happens if player and wall collide. In the example the platform is spawned as an entity at the position taken from Tiled, but where is it transformed to a platform, meaning the player can't run through it?
Thanks for your help!

**Reply by stingerhorse** [reply]
Can't figure it out either.

**Reply by AlmasB** [reply]
>  but where is it transformed to a platform

In the Mario example above, it's converted from a Tiled object to an FXGL entity [in the factory](https://github.com/AlmasB/FXGLGames/blob/d00f131f6911e5e865d540886bfab138167ca004/Mario/src/main/java/com/almasb/fxglgames/platformer/PlatformerFactory.java#L41). The factory is the only place where entities get created.

**Reply by fleggs** [reply]
Yeah I get that, but where do I say that the wall is impenetrable by the player and other entities?

**Reply by AlmasB** [reply]
In the same method. More specifically:
```
    @Spawns("platform")
    public Entity newPlatform(SpawnData data) {
        return entityBuilder(data)
                .type(PLATFORM)
                .bbox(new HitBox(BoundingShape.box(data.<Integer>get("width"), data.<Integer>get("height"))))       /// 1.
                .with(new PhysicsComponent())                                                                       /// 2.
                .build();
    }
```

Line `1.` adds a bounding box around the platform using the Tiled object width and height. Line `2.` adds a physics component, so that the platform entity is controlled by the physics world. This creates a rigid body in the physics world, which, in turn, makes the platform "impenetrable".

Note that other entities and the player must also have rigid bodies, using the same principles.

**Reply by AlmasB** [reply]
Another thing of note is that transformations of all entities with `PhysicsComponent` must only be modified via the physics world, as it has exclusive access to them. So instead of setX/setY, one would call `getComponent(PhysicsComponent.class).setLinearVelocity(...)`.

**Reply by fleggs** [reply]
Thanks that helped a lot. I added the PhysicsComponent to the entities and switched from viewWithBBox() to bbox(). In my AnimationComponent class, which is the same as the "PlayerComponent" class in the platformer example, I now changed from speedx and speedy to entity.getComponent(PhysicsComponent.class).isMoving() . The game starts, but the player isn't able to move. I added the class in the following for a better overview. I'm sorry for asking as much but I'm a beginner so any help is appreciated.
`
package com.almasb.game;

import com.almasb.fxgl.core.math.FXGLMath;
import com.almasb.fxgl.dsl.FXGL;
import com.almasb.fxgl.entity.component.Component;
import com.almasb.fxgl.texture.AnimatedTexture;
import com.almasb.fxgl.texture.AnimationChannel;
import com.almasb.fxgl.physics.PhysicsComponent;
import javafx.geometry.Point2D;
import javafx.util.Duration;

public class AnimationComponent extends Component {

    public int speedx = 0;
    public int speedy = 0;
    public boolean attack = false;
    private AnimatedTexture texture;
    private AnimationChannel animIdle, animWalk, animAttack;

    public AnimationComponent() {
        animIdle = new AnimationChannel(FXGL.image("PlayerIdle.png"), 1, 30, 50, Duration.seconds(1), 0, 0);
        animWalk = new AnimationChannel(FXGL.image("PlayerWalk.png"), 6, 30, 50, Duration.seconds(0.8), 0, 5);
        animAttack = new AnimationChannel(FXGL.image("PlayerWalk.png"), 3, 30, 50, Duration.seconds(0.3), 0, 2);

        texture = new AnimatedTexture(animIdle);
    }

    @Override
    public void onAdded() {
        entity.getTransformComponent().setScaleOrigin(new Point2D(16, 21));
        entity.getViewComponent().addChild(texture);
    }

    @Override
    public void onUpdate(double tpf) {

        if (entity.getComponent(PhysicsComponent.class).isMoving() == true) {
            if (texture.getAnimationChannel() != animWalk) {
                texture.loopAnimationChannel(animWalk);
            }
        }

        if (entity.getComponent(PhysicsComponent.class).isMoving() == false) {
            if (texture.getAnimationChannel() != animIdle) {
                texture.loopAnimationChannel(animIdle);
            }
        }

        if (attack == false && entity.getComponent(PhysicsComponent.class).isMoving() == true) {
            if (texture.getAnimationChannel() != animWalk) {
                texture.loopAnimationChannel(animWalk);
            }
        }

        if (attack == false && entity.getComponent(PhysicsComponent.class).isMoving() == false) {
            if (texture.getAnimationChannel() != animIdle) {
                texture.loopAnimationChannel(animIdle);
            }
        }

        if (attack == true && entity.getComponent(PhysicsComponent.class).isMoving() == true) {
            if (texture.getAnimationChannel() != animAttack) {
                texture.playAnimationChannel(animAttack);
            }
        }

        if (attack == true && entity.getComponent(PhysicsComponent.class).isMoving() == false) {
            if (texture.getAnimationChannel() != animAttack) {
                texture.playAnimationChannel(animAttack);
            }
        }

        attack = false;

    }

    public void moveRight() {
        entity.getComponent(PhysicsComponent.class).setVelocityX(150);
        getEntity().setScaleX(1);
    }

    public void moveLeft() {
        entity.getComponent(PhysicsComponent.class).setVelocityX(-150);
        getEntity().setScaleX(-1);
    }

    public void moveUp() {
        entity.getComponent(PhysicsComponent.class).setVelocityY(-150);
    }

    public void moveDown() {
        entity.getComponent(PhysicsComponent.class).setVelocityY(150);
    }

    public void doAttack() {
        attack = true;
    }
}
`

**Reply by lhDream** [reply]
@fleggs I think you can try this.
~~~java
    @Spawns("platform")
    public Entity newPlatform(SpawnData data) {
        PhysicsComponent physics = new PhysicsComponent();
        physics.setBodyType(BodyType.DYNAMIC);
        FixtureDef fd = new FixtureDef();
        fd.setDensity(0.7f);
        fd.setRestitution(0.3f);
        physics.setFixtureDef(fd);
        return entityBuilder(data)
                .type(PLATFORM)
                .bbox(new HitBox(BoundingShape.box(data.<Integer>get("width"), data.<Integer>get("height"))))     
                .with(physics)                                                                   
                .build();
    }

~~~

**Reply by fleggs** [reply]
Thanks for the help, but I think this doesn't work for me, since I want to create a Top-Down game. So you can move up, down, left and right and you are not able to jump. So the dynamic body type of the platform and the player is not the right in my case. I tried the kinematic body type on the player and you can move the player now how I wanted it to be.

But I still can move through walls, eventhough the walls and the player have the PhysicsComponent integrated.

I tried to use "viewWithBbox" instead of just "bbox" and gave the walls a colour, so I could see if the walls are spawned correctly. And they are. But I can walk through them anyways. Any help with that? Should I attach my code of the spawning class?

**Reply by lhDream** [reply]
@fleggs In that case, maybe you can try this.
~~~java
    @Spawns(PLAYER)
    public Entity newPlayer(SpawnData data) {
        var texture = texture("player.png");
        var topDownMoveComponent = new TopDownMoveComponent(200);
        var build = entityBuilder(data)
                .type(EntityType.PLAYER)
                .at(getAppWidth() / 2 - texture.getWidth() / 2, getAppHeight() / 2 - texture.getHeight() / 2)
                .viewWithBBox(texture)
                .collidable()
                .with(topDownMoveComponent)
                .build();
        build.setOnActive(()->topDownMoveComponent.collidables(WALL));
        return build;
    }
~~~

~~~java
    @Override
    protected void initInput() {
        onKey(KeyCode.W,()->{
            player.getComponent(TopDownMoveComponent.class).moveUp();
        });
    }
~~~

**Reply by fleggs** [reply]
Thanks a lot, I managed to find a solution for the problem. I now want to position the hitbox of the player on the feet of the character, right now it is at the head of the texture of the player. Heres how it looks in game, the blue box is the hitbox. Is it possible to position it on the bottom of the player texture?

![Unbenannt](https://user-images.githubusercontent.com/102703005/164091483-30e20fde-c620-4f10-8007-4bf7bb06fa99.JPG)

**Reply by AlmasB** [reply]
When constructing a `HitBox`, you can supply an extra parameter - offset - that changes where the box's origin is located.

**Reply by fleggs** [reply]
Yeah I got it, thanks!

**Reply by fleggs** [reply]
Is it possible to add a FollowComponent to an entity with a PhysicsComponent? Because the FollowComponent comes with an own speed which I think isn't accepted by the PhysicsComponent, so the entity just doesn't move. Is it possible to make an entity follow the player, but also participating in the PhysicsWorld?

**Reply by stingerhorse** [reply]
I am struggling with it too. I tried to set the velocity from the PhysicsComponent of the entity to the speed of the FollowComponent:

`entity.getComponent(PhysicsComponent.class).setVelocityX(entity.getComponent(FollowComponent.class).getSpeed());
entity.getComponent(PhysicsComponent.class).setVelocityY(entity.getComponent(FollowComponent.class).getSpeed());`

Thought it would be a good idea but it's not working :-/

**Reply by fleggs** [reply]
Anybody else has an idea? I appreciate every answer.

**Reply by lhDream** [reply]
You can talk about the functions you want to achieve, and there may be alternative solutions.

**Reply by AlmasB** [reply]
> Is it possible to add a FollowComponent to an entity with a PhysicsComponent?

There is no built-in component that does this. However, one can easily build something similar for `PhysicsComponent`.

Let `e1` and `e2` be entities with `PhysicsComponent`, such that `e1` follows `e2`. Given we know positions of both entities, all that remains is to use these positions to calculate `e1` velocity and apply via the component.
